File: control_server_cpp/src/expire_manager.cpp
#include "expire_manager.h"

using namespace zhicloud;
using namespace control_server;
using namespace util;


const uint32_t ExpireManager::_tresource_type_host(0);
const uint32_t ExpireManager::_expire_type_found(0);
const uint32_t ExpireManager::_expire_type_lost(1);

ExpireManager::ExpireManager(const string& logger_name)
{
    _logger = getLogger(logger_name);
}

ExpireManager::~ExpireManager()
{
    //dtor
}

bool ExpireManager::start(const string& resource_name)
{
    lock_type lock(_mutex);

    string found_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_found).str();
    _expire[found_name] = move(map<UUID_TYPE,uint32_t>());

    string lost_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_lost).str();
    _expire[found_name] = move(map<UUID_TYPE,uint32_t>());

    return true;
}

bool ExpireManager::updateFound(const string& resource_name, const set<UUID_TYPE>& id_list)
{
    string expire_key = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_found).str();
   return update(expire_key,id_list);
}

bool ExpireManager::updateLost(const string& resource_name, const set<UUID_TYPE>& id_list)
{
    string expire_key = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_lost).str();
    return update(expire_key,id_list);
}

bool ExpireManager::update(const string& target_name,const set<UUID_TYPE>& id_list)
{
    lock_type lock(_mutex);
    auto target_iter = _expire.find(target_name);
    if(target_iter != _expire.end() )
    {
        map<UUID_TYPE,uint32_t>& target = target_iter->second;
        for(auto& id : id_list)
        {
            auto found_iter = target.find(id);
            if(found_iter == target.end())
                target.emplace(id,1);//new
            else
                found_iter->second +=1;
        }

        auto del_iter = target.begin();
        for(;del_iter!= target.end();)
        {
            if(id_list.count(del_iter->first) == 0)
            {
                target.erase(del_iter++);
            }else{
                del_iter++;
            }
        }
    }

    return false;
}

bool ExpireManager::isFound(const string& resource_name,const UUID_TYPE& uuid,const uint32_t & max_expire)
{
    string found_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_found).str();
    return check(found_name,uuid,max_expire);

}
bool  ExpireManager::isLost(const string& resource_name,const UUID_TYPE& uuid,const uint32_t & max_expire)
{
    string lost_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_lost).str();
    return check(lost_name,uuid,max_expire);
}

bool ExpireManager::check(const string& target_name,const UUID_TYPE& uuid,const uint32_t & max_expire)
{
    lock_type lock(_mutex);

    auto target_iter = _expire.find(target_name);
    if(target_iter != _expire.end() )
    {
        auto iter = target_iter->second.find(uuid);
        if(iter ==  target_iter->second.end())
            return false;

        iter->second +=1;
        if(iter->second < max_expire )
            return false;

        target_iter->second.erase(iter);
             return true;
    }
    return false;
}

bool  ExpireManager::finish(const string& resource_name)
{
    lock_type lock(_mutex);
    string found_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_found).str();
    auto found_iter = _expire.find(found_name);
    if(found_iter != _expire.end() )
        _expire.erase(found_iter);

    string lost_name = (boost::format("%d_%s_%d") %_tresource_type_host %resource_name %_expire_type_lost).str();
    auto lost_iter = _expire.find(lost_name);
    if(lost_iter != _expire.end() )
        _expire.erase(lost_iter);

    return true;
}

File: control_server/expire_manager.py
#!/usr/bin/python
import threading

class ExpireManager(object):
    
    resource_type_host = 0
    
    expire_type_found = 0
    expire_type_lost = 1
    
    def __init__(self):
        self.lock = threading.RLock()
        ##key = name:[resource_type]_[resource_name]_[expire_type],etc: "0_ncxxx_1"
        ##value = map<id, counter>
        self.expire = {}

    def start(self, resource_name):
        with self.lock:
            found_name = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_found)
            self.expire[ found_name ] = {}
            lost_name = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_lost)
            self.expire[ lost_name ] = {}
            return True            

    def updateFound(self, resource_name, id_list):
        with self.lock:
            expire_key = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_found)
            if not self.expire.has_key(expire_key):
                return False
            remove_list = []
            for object_id in self.expire[expire_key].keys():
                if object_id not in id_list:
                    remove_list.append(object_id)
            if 0 != len(remove_list):
                for object_id in remove_list:
                    del self.expire[expire_key][object_id]
            for object_id in id_list:
                if self.expire[expire_key].has_key(object_id):
                    ##increase counter
                    self.expire[expire_key][object_id] += 1
                else:
                    ##new counter
                    self.expire[expire_key][object_id] = 1

    def updateLost(self, resource_name, id_list):
        with self.lock:
            expire_key = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_lost)
            if not self.expire.has_key(expire_key):
                return False
            remove_list = []
            for object_id in self.expire[expire_key].keys():
                if object_id not in id_list:
                    remove_list.append(object_id)
            if 0 != len(remove_list):
                for object_id in remove_list:
                    del self.expire[expire_key][object_id]
            for object_id in id_list:
                if self.expire[expire_key].has_key(object_id):
                    ##increase counter
                    self.expire[expire_key][object_id] += 1
                else:
                    ##new counter
                    self.expire[expire_key][object_id] = 1
                    
    def isFound(self, resource_name, object_id, max_expire):
        with self.lock:
            expire_key = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_found)
            if not self.expire.has_key(expire_key):
                return False
            ##increase
            if not self.expire[expire_key].has_key(object_id):
                return False
            ##check
            self.expire[expire_key][object_id] += 1
            if self.expire[expire_key][object_id] < max_expire:
                ##not expire
                return False
            ##expired&clear
            del self.expire[expire_key][object_id]
            return True

    def isLost(self, resource_name, object_id, max_expire):
        with self.lock:
            expire_key = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_lost)
            if not self.expire.has_key(expire_key):
                return False
            ##increase
            if not self.expire[expire_key].has_key(object_id):
                return False
            ##check
            self.expire[expire_key][object_id] += 1
            if self.expire[expire_key][object_id] < max_expire:
                ##not expire
                return False
            ##expired&clear
            del self.expire[expire_key][object_id]
            return True

    def finish(self, resource_name):
        with self.lock:
            found_name = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_found)
            if self.expire.has_key(found_name):
                del self.expire[found_name]
            lost_name = "%d_%s_%d"%(ExpireManager.resource_type_host, resource_name, ExpireManager.expire_type_lost)
            if self.expire.has_key(lost_name):
                del self.expire[lost_name]
            return True            
    
        
